### DISCUSSION TITLE
Python implemention of `foxglove/rosmsg-serialization`

### DISCUSSION THREAD

**Discussion by muety** [foxglove/community#945] [General]
I am looking for a ROS-free (de-)serializer for common ROS messages, that allows me to encode and decode `sensor_msgs` and others without having to have an actual ROS distribution installed (esp. since ROS 1.x doesn't work on recent OS). Essentially a Python implementation of https://www.npmjs.com/package/@foxglove/rosmsg-serialization. Is there anything like that already around or perhaps planned?

**Comment by jtbandes** [comment]
You can take a look at the `mcap-ros1-support` package which vendors some of the dynamic deserialization code from ROS 1 so it can be run outside a ROS environment: https://github.com/foxglove/mcap/blob/734c6433b4e3ab0df403a4c41b2e6b853d6586ec/python/mcap-ros1-support/mcap_ros1/decoder.py#L4-L10

**Reply by muety** [reply]
Thanks! I'm not exactly sure how to use this, though. I read through the code and came up with this hacks solution, but I think that's not the proper way to go:

```python
schema1: Schema = Schema(id=123, name='sensor_msgs/Image, encoding=SchemaEncoding.ROS1, data=open('/tmp/Image.msg', 'r').read().encode('utf-8'))
d1: Decoder = DecoderFactory().decoder_for('ros1', schema1)

def my_callback(data: bytes):
    return d1(data)
```

For this to work, I created `/tmp/Image.msg`  (from [`Image.msg`](https://docs.ros.org/en/noetic/api/sensor_msgs/html/msg/Image.html)) and manually embedded the fields from [`Header.msg`](https://docs.ros.org/en/noetic/api/std_msgs/html/msg/Header.html), because I didn't know how to dynamically include that.

I'm sure there's a better way to do this. Essentially I want to be able to to sth. like

```python
decoder = DecoderFactory().decoder_for('ros1', magically_get_me_a_ros1_schema_for_type('sensor_msgs/Image'))
```

I found [`msg_loader.py`](https://github.com/foxglove/mcap/blob/main/python/mcap-ros1-support/mcap_ros1/_vendor/genmsg/msg_loader.py), which might be useful, but wasn't sure how to utilize that either.

Could you please point me into the right direction here?

**Reply by muety** [reply]
Any tips you could give me on this end?

**Reply by jtbandes** [reply]
I'm not sure what I can help with exactly, but you will definitely need access to the `.msg` files in some form. In MCAP they are embedded in the Schema records like this: https://mcap.dev/spec/registry#ros1msg

**Reply by muety** [reply]
Access to the `.msg` files is not a problem. I'm only wondering what the correct APIs / functions are to load them and use them for parsing ROS messages. Could you give a brief code example for my use case?
